---
title: "BBSRC_ARC"
author: "Simon Babayan & Wei Liu"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, echo=F, message=F, warning=FALSE}
rm(list=ls())
options(stringsAsFactors = FALSE)

# samples preparation
library("readr")
library("readxl")
library("dplyr")
library("plyr")
library("ensembldb")
library("biomaRt")
library("tximport")
library("edgeR")
library("DESeq2")
library("Rtsne")
library("ggplot2")
library("extrafont")
library("maSigPro")
library("MASS")
library("RColorBrewer")
library("gplots")
library("reshape2")
library("stringr")
library("scales")
library("glmnet")
library("ComplexHeatmap")
library("tidyr")
library("VennDiagram")

```

<!-- prepare the RNA-seq sample files -->

# RNAseq samples samplesheet
```{r, sample preparation}
# user's RNAseq reads count directory. this example is from Kallisto
base_dir <- "~/Dropbox/wei_glasgow_work/kallisto_tximport/full_kallisto_data/" 
sample_id <- list.files(path=base_dir)
kal_dirs <- as.data.frame(t(sapply(sample_id, function(id) c(id,file.path(base_dir,id)))))
colnames(kal_dirs) <- c("sample","path")
head(kal_dirs)

# samplesheet directory 
RNA_samplesheet <- read_excel("~/Dropbox/sheep_arc/data/read_in/NGSSampleForm_filltered.xlsx")
RNA_samplesheet <- RNA_samplesheet[,-1]
colnames(RNA_samplesheet) <- c("sample","group_week", "concentration")
RNA_samplesheet$group <- unlist(lapply(strsplit(gsub('([[:upper:]])', ' \\1', as.character(RNA_samplesheet$group_week))," "), "[", 2))
RNA_samplesheet$week <- unlist(lapply(strsplit(gsub('([[:upper:]])', ' \\1', as.character(RNA_samplesheet$group_week))," "), "[", 3))
RNA_samplesheet <- RNA_samplesheet[,c("sample", "group", "week","concentration")]
sample_list <- cbind(RNA_samplesheet[match(kal_dirs[,1],RNA_samplesheet$sample),], kal_dirs[,2])
sample_list[,5] <- as.character(sample_list[,5])
colnames(sample_list)[5] <- "path"
head(RNA_samplesheet)


```

#Select RNA-seq samples acrossing time and treatment
# RNAseq samples metadata
```{r, echo=T}
# week!="Week9"&week!="Lymph"; group=="Group3"|group=="Group4"
# We don't use week9 and lymph-nodes samples
subset_samples <- dplyr::filter(sample_list,week!="Week9"&week!="Lymph" )
#save(subset_samples, file="data/res/sample_table_w1_7.RData")
files <- file.path(subset_samples$path, "abundance_change.tsv")
names(files) <- subset_samples[,"sample"]
all(file.exists(files))

head(subset_samples)


```


# Ensembl transcript ID to gene sysmbol mapping
```{r find gene name for each transcript ID}
# Mapping transcript ID with Gene name
# mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL",
#                          dataset = 	"oaries_gene_ensembl",#oaries_gene_ensembl",
#                          host = 'ensembl.org')
#tx2gene <- getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "external_gene_name"), mart = mart)
# ensembl=useMart("ensembl")
# a=listDatasets(ensembl)
# listMarts(host="ensembl.org")
# write.table(a, "list_ensembl.csv")
#save(tx2gene, file="tx2gene.RData")
#write.table(tx2gene, file="tx2gene.csv", row.names = F)

# Here we load the pre-saved mapping table. user can use biomart to generate own gene table
load("~/Dropbox/sheep_arc/tx2gene.RData")
tx2gene_name <- tx2gene[,c(1,3)]
head(tx2gene_name,10)
```
# import count matrix from tximport
```{r, message=FALSE}
# Import and summarize transcript-level abundance estimates for transcript- and gene-level analysis--tximport
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_name)
# rows -- genes & cols -- samples
# the first line in txi$counts is the sum of counts for the transcripts which have no gene name.
print(txi$counts[1:5,1:10])
txi$counts <- txi$counts[-1,]
txi$length <- txi$length[-1,]
txi$abundance<- txi$abundance[-1,]
```

## Normalisation
### edgeR
```{r edgeR--from tximport}
cts <- txi$counts
normMat <- txi$length
normMat <- normMat/exp(rowMeans(log(normMat))) # average transcript length across samples
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y <- DGEList(cts)
y$offset <- t(t(log(normMat)) + o)
all.equal(rownames(y$samples), subset_samples[,"sample"] )
y$samples$group <- subset_samples[,"group"]
y$samples$week <- subset_samples[,"week"]
y$samples <- y$samples[,c(1,4,2,3)]
```

```{r edgeR--filtering low counts and generate normalised edgeR read counts table}
y <- calcNormFactors(y)
#the count for each sample is > 5 and across at least 7 samples for each gene.
keep <- rowSums(cpm(y)>5) >= 7
#  recalculate the library sizes of the DGEList object y
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
edgeR_norm <- cpm(y)
dim(edgeR_norm)
hist(edgeR_norm[edgeR_norm<=500], breaks=50, col = "grey", main = "Histogram of normalised counts by edgeR")
```

### DESeq2
```{r, normalised DESeq2 reads count matrix}
sampleTable <- data.frame(condition = factor(colnames(txi$counts)))
rownames(sampleTable) <- colnames(txi$counts)
dds <- DESeqDataSetFromTximport(txi, sampleTable, ~condition)
dds <- estimateSizeFactors(dds)
idx <- rowSums( counts(dds, normalized=TRUE) >= 5 ) >= 2
dds <- dds[idx,]
DESeq2_norm <- counts(dds, normalized=TRUE)
dim(DESeq2_norm)
hist(DESeq2_norm[DESeq2_norm<=1000], breaks=50, col = "grey", main = "Histogram of normalised counts by DESeq2")
```

## Dimensionality reduction
```{r, ggplot_dimRedu}
# The ellipse is 95% (default level) 2D confidence interval for RNA-seq samples which are assumed drawn from X distribution.The default distribution is multivariate t-distribution. alpha is the percentage of transparency.
ggplot_dimRedu <- function(input_data, sample_size, X, mtd){
 ggplot() +
  stat_ellipse(type="t",geom="polygon",data=input_data, alpha=0.15, aes(x=V1,y=V2,fill=week,group=week))+
  geom_point(data=input_data,aes(x=V1,y=V2,colour=week, shape=group),size=4) +# add the point markers
  scale_shape_manual(values = c(2,17,1,16))+
  # use sample name as label to find the outliers
  #geom_point()+
  #geom_text(data=dimRedu,aes(x=V1, y=V2,label=name),hjust=0, vjust=0)+
  theme_bw() + #ggtitle(paste(X,mtd,"for",sample_size,"RNA-seq samples", sep=" "))+
   theme(
        axis.ticks = element_blank(),  # remove axis ticks
        axis.title.x = element_text(size=18), # remove x-axis labels
        axis.title.y = element_text(size=18), # remove y-axis labels
        panel.background = element_blank(),
        panel.grid.major = element_blank(),  #remove major-grid labels
        panel.grid.minor = element_blank(),  #remove minor-grid labels
        plot.background = element_blank())
}
```


```{r, PCA or MDS / edgeR or DESeq2}
# choose between edgeR and DESeq2 normalised reads count matrix. 
X <- "DESeq2" ##
X_norm <- get(paste0(X,"_norm"))
print(X_norm[1:5,1:7])
# choose between Dimensionality reduction methods: PCA, MDS or tSNE
mtd <- "tSNE"
```

### PCA, MDS & tSNE

```{r, PCA, MDS and t-SNE}
#ntop=10000
#rv = rowVars(X_norm)
#select = order(rv, decreasing=TRUE)[seq_len(ntop)]
#pca = prcomp(t(X_norm[select,]))

if (mtd=="PCA") {
    X_pca <- prcomp(X_norm, center = TRUE, scale. = TRUE)
  # summary(X_pca)
  #plot(X_pca, type="l")
  dimRedu <- as.data.frame(X_pca$rotation[,1:2])
} else if (mtd=="MDS") {
    dist_matrix <- dist(t(X_norm))
   fit <- cmdscale(dist_matrix,eig=TRUE, k=2) # k is the number of dim
   dimRedu<- as.data.frame(fit$points)
} else if (mtd == "tSNE") {
  rtsne_in <- t(as.matrix(unique(X_norm)))
   rtsne_out_2 <- Rtsne(rtsne_in, dims=2)
   dimRedu <- as.data.frame(rtsne_out_2$Y)
}
```


```{r, ggplot}

# define the Dimensionality reduction plot labels here
colnames(dimRedu) <- c("V1","V2")
dimRedu$group<- subset_samples$group
dimRedu$group[dimRedu$group=="Group1"] <- "3mo_vax"
dimRedu$group[dimRedu$group=="Group2"] <- "3mo_ctrl"
dimRedu$group[dimRedu$group=="Group3"] <- "6mo_vax"
dimRedu$group[dimRedu$group=="Group4"] <- "6mo_ctrl"
dimRedu$week <- subset_samples$week
dimRedu$name <- rownames(dimRedu)
dimRedu$week[dimRedu$week=="Week9"] <- "Abomasum"
dimRedu$week[dimRedu$week=="Lymph"] <- "Lymph Node"

ggplot_dimRedu(dimRedu, sample_size="180", X, mtd)

```




## FEC & Worm Count
### Worm Count
```{r, worm count in each aging & treatment group}
# read parasitological data
allTraits <- read_xlsx("~/Dropbox/sheep_arc/data/read_in/Complete data set Tcirc AHRC 010617.xlsx",sheet=1, col_names = T)[1:62,]
dim(allTraits)
datTraits_worm <- allTraits[,c("Animal_ID","Group","Treatment","Total_Worms")]
source("~/Dropbox/sheep_arc/ztheme.R")
cols <- c("orangered3","lightpink2","royalblue3","skyblue1")
datTraits_worm$Group <- as.factor(datTraits_worm$Group)
ggplot(datTraits_worm, aes(x=Group,y=Total_Worms) )+
  geom_boxplot(aes(fill=Group),alpha=.5, outlier.shape = NA)+
  geom_jitter(aes(color=Group),size=3,alpha=.4)+
  scale_fill_manual(values=cols)+
  scale_color_manual(values=cols)+
  #scale_y_continuous(breaks=seq(0,1,.1), labels=scales::percent)+
  guides(fill=FALSE,color=FALSE)+
  labs(x="\nLamb group",y="Total Worm Count \n")+
  scale_x_discrete(breaks=c("1","2","3","4"),labels=c("3mo_Vax", "3mo_Ctrl", "6mo_Vax", "6mo_Ctrl"))+
  z_theme()

```


### FEC
## AUC FEC per aging & treatment group
```{r}
datTraits_AUC <-  cbind(allTraits[,c("Animal_ID","Group","Treatment")],allTraits[,grepl("AUC", names(allTraits))])

## Time series AUC
group_matrix <- matrix(0,4,ncol(datTraits_AUC)-3)
group_sd <- group_matrix
for(i in 1:4) {
  filter_group <- dplyr::filter(datTraits_AUC, Group == i)[,-c(1:3)]
  group_matrix[i,] <- colMeans(filter_group)
  group_sd[i,] <- apply(filter_group,2,sd)
}
plot_colors <- c("red","indianred2","blue","royalblue1")
matplot(t(group_matrix),
        col=plot_colors, type="o", xaxt = "n", ylab = "FEC", xlab="Days post 1st vaccination",
        lwd=3, lty=c(1,2,1,2), main="AUC Faecal Egg Count Time Series (Mean)")
axis(1, at=1:length(colnames(datTraits_AUC)[-c(1:3)]), labels=colnames(datTraits_AUC)[-c(1:3)])
legend(1, 10000, legend=c("1: 3M_Vax", "2: 3M_Ctrl", "3: 6M_Vax", "4: 6M_Ctrl"),
       col=plot_colors, lty=c(1,2,1,2), cex=0.8)

```


# Cumulative FEC boxplot per group
```{r}
cols <- c("orangered3","lightpink2","royalblue3","skyblue1")
datTraits_AUC$Group <- as.factor(datTraits_AUC$Group)
ggplot(datTraits_AUC, aes(x=Group,y=AUCDay95) )+
  geom_boxplot(aes(fill=Group),alpha=.5, outlier.shape = NA)+
  geom_jitter(aes(color=Group),size=3,alpha=.4)+
  scale_fill_manual(values=cols)+
  scale_color_manual(values=cols)+
  #scale_y_continuous(breaks=seq(0,1,.1), labels=scales::percent)+
  guides(fill=FALSE,color=FALSE)+
  labs(x="\nLamb group",y="Cumulative FEC\n")+
  scale_x_discrete(breaks=c("1","2","3","4"),labels=c("3mo_Vax", "3mo_Ctrl", "6mo_Vax", "6mo_Ctrl"))+
  z_theme()
```

## Biopsy samples Analysis: dynamic pairwise comparison
### maSigPro
```{r, maSigPro}
load("~/Dropbox/sheep_arc/data/res/DESeq2_norm.RData")
dim(DESeq2_norm)
# sample table for maSigPro
group_num <- c("Group1", "Group2", "Group3", "Group4")
week_num <- c("Week1", "Week2", "Week3", "Week4", "Week5", "Week7")
maSig_samples <- subset_samples[,c("sample", "group", "week")]
for (i in 1:4) {
  maSig_samples<- cbind(maSig_samples, as.numeric(maSig_samples$group==group_num[i]))
}
colnames(maSig_samples)[4:7] <- group_num
maSig_samples$replicates <- maSig_samples$Group1
ind <- 1
for (i in 1:length(group_num)) {
  for(j in 1:length(week_num)) {
  maSig_samples$replicates[c(which(maSig_samples$group==group_num[i]&maSig_samples$week==week_num[j]))]=ind
  ind = ind+1
  }
}
test_re = 8
subset(maSig_samples, maSig_samples$replicates==test_re)$replicates == test_re
maSig_samples$time<- as.numeric(substring(maSig_samples$week,5))
rownames(maSig_samples) <- maSig_samples$sample

# Select Groups
sel_group <- "1vs3"
maSig_sub <- subset(maSig_samples, Group1!=0 | Group3!=0)
maSig_sub_table <- maSig_sub[,c("time", "replicates","Group1","Group3")]
X_sub <- round(DESeq2_norm[,match(rownames(maSig_sub_table), colnames(DESeq2_norm))]) # simulate counts
dim(X_sub)
#head(X_sub)

###Run maSigpro

# d <- make.design.matrix(maSig_sub_table)
# NBp <- p.vector(X_sub, d, counts=TRUE)
# NBt <- T.fit(NBp)
# save(NBp, NBt, file=paste0("masigpro/NB_",sel_group,".RData"))
load(paste0("~/Dropbox/sheep_arc/masigpro/NB_",sel_group,".RData"))
get<-get.siggenes(NBt, vars="all",  rsq = 0.4)
get$summary
a = see.genes(get$sig.genes, k =7)
save(get, file = paste0("masigpro/get_",sel_group,".RData"))
gene_cluster <- as.data.frame(a$cut)
colnames(gene_cluster) <- "cluster"
#save(gene_cluster, file=paste0("masigpro/gene_cluster_",sel_group,".RData"))
#write.table(gene_cluster, file=paste0("masigpro/gene_cluster_",sel_group,".csv"))

```



## plot masigPro results 
```{r, }
mean_matrix <- matrix(0, nrow(DESeq2_norm),24)
rownames(mean_matrix) <- rownames(DESeq2_norm)
colnames(mean_matrix) <- colnames(mean_matrix, do.NULL = FALSE)
for (i in 1:24) {
  mean_matrix[,i]= rowMeans(DESeq2_norm[,match(rownames(subset(maSig_samples, replicates==i)),colnames(DESeq2_norm))])
  colnames(mean_matrix)[i] <- paste(substr(rownames(subset(maSig_samples, replicates==i))[1],7,8),substr(rownames(subset(maSig_samples, replicates==i))[1],5,6))
}
mean_matrix <- mean_matrix[,order(colnames(mean_matrix))]

load(paste0("~/Dropbox/sheep_arc/masigpro/gene_cluster_",sel_group,".RData"))
gene_cluster <- as.data.frame(cbind(rownames(gene_cluster), gene_cluster$cluster))
colnames(gene_cluster) <- c("gene", "cluster")
gene_list <- gene_cluster[order(gene_cluster$cluster),]
gene_list <- split(gene_list, f = gene_list$cluster)
mean_matrix_sub <- mean_matrix[,substr(colnames(mean_matrix),4,5)=="g1"|
                                 substr(colnames(mean_matrix),4,5)=="g3"]


#plot
for (ind_cl in 1:7) {
  map_matrix <- log2(mean_matrix_sub[match(gene_list[[ind_cl ]]$gene,rownames(mean_matrix_sub)),])
  map_matrix[which(is.infinite(map_matrix))] <- 0
  map_matrix <- t(scale(t(map_matrix)))
  print(min(map_matrix))
  print(max(map_matrix))

  hmcol = colorRampPalette(brewer.pal(9, "RdBu"))(100)
  myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")), space="Lab")
  png(filename=paste("~/Dropbox/sheep_arc/masigpro/",sel_group,"_heatmap_scale/",ind_cl ,"_heatmap.png", sep=""))
  heatmap.2(map_matrix,scale="none", col = myPalette, trace="none",
            cexCol=0.9,cexRow=0.5,density.info="density", breaks = seq(-2.67, 3, length.out = 200),
            Rowv=T, Colv=F)
  dev.off()

}

```


```{r, bar plot of masigpro DE gene numbers}
source("ztheme.R")
ma_gene_cluster <- read.csv("~/Dropbox/sheep_arc/masigpro/gene_cluster_1vs2.csv", sep=" ")
df <- as.data.frame(cbind(rownames(ma_gene_cluster), ma_gene_cluster$cluster,"3V vs 3C"))
ma_gene_cluster <- read.csv("~/Dropbox/sheep_arc/masigpro/gene_cluster_1vs3.csv", sep=" ")
df <- as.data.frame(rbind(df, cbind(rownames(ma_gene_cluster), ma_gene_cluster$cluster, "3V vs 6V")))
ma_gene_cluster <- read.csv("~/Dropbox/sheep_arc/masigpro/gene_cluster_3vs4.csv", sep=" ")
df <- as.data.frame(rbind(df, cbind(rownames(ma_gene_cluster), ma_gene_cluster$cluster,"6V vs 6C")))
ma_gene_cluster <- read.csv("~/Dropbox/sheep_arc/masigpro/gene_cluster_2vs4.csv", sep=" ")
df <- as.data.frame(rbind(df, cbind(rownames(ma_gene_cluster), ma_gene_cluster$cluster,"3C vs 6C")))
colnames(df) <- c("Gene", "Cluster", "Group")
df$Group <- factor(df$Group, levels=c("3C vs 6C", "3V vs 3C", "6V vs 6C","3V vs 6V" ))
 nrow(subset(df, Group=="3V  vs  6V"))

ggplot(df, aes(Cluster, fill = Cluster))+
    theme_bw()+
    labs(y = "The number of DE Genes in each cluster", x = "Cluster")+
    geom_bar()+
    scale_fill_brewer(palette = "Set2")+
    facet_grid(Group ~ .)+
    theme(legend.text=element_text(size=12),legend.position = "none")+
    theme(strip.text.y = element_text(size = 12, face="bold"))+
    theme(axis.title.x=element_text(size=12, vjust=0, face="bold")) +
    theme(axis.title.y=element_text(size=12, vjust=1.25, face="bold"))

```



## Venn Diagram of 4 groups
```{r}
library("VennDiagram")
library("gplots")
  G_1vs2 <- subset(df, Group=="3V vs 3C")$Gene
  G_1vs3 <- subset(df, Group=="3V vs 6V")$Gene
  G_2vs4 <- subset(df, Group=="3C vs 6C")$Gene
  G_3vs4 <- subset(df, Group=="6V vs 6C")$Gene
  venn(list(GrpA=G_2vs4,GrpB=G_1vs2, GrpC=G_3vs4, GrpD=G_1vs3))


```


```{r, match ipa immune database}
imm_list <- read.xlsx("~/Dropbox/sheep_arc/data/read_in/immune_pathway_list.xls",1, header=T)
imm_list <- imm_list[,2:4]
colnames(imm_list) <- imm_list[1,]
imm_list <- unique(imm_list[-1,])

pathway_heatmap <- NULL
file_list <- c("2vs4_3","1vs2_3","3vs4_2","1vs3_2")
for (i_file in file_list) {
pathway_week <- read.xlsx(paste0("~/Dropbox/sheep_arc/masigpro/ingenuity/pathway_heatmap_v2/ipa_",i_file,".xls"),1)
pathway_week <- pathway_week[-1, c(1,2,5)]
colnames(pathway_week) <- c("pathway", "log_p", "gene")
pathway_index <- na.omit(match(imm_list$Name,pathway_week$pathway))
pathway_group <- pathway_week[pathway_index,]
pathway_group$group <- substring(i_file,1,4)
pathway_heatmap <- rbind(pathway_heatmap, pathway_group)
}
```


# Pathway heatmap with log_p 
```{r, pathway heatmap}
read_ipa_table <- subset(pathway_heatmap, log_p >= 1.3)[,-3]
read_ipa_table$log_p <- as.numeric(read_ipa_table$log_p)
read_ipa_table$pathway <- as.factor(read_ipa_table$pathway)
read_ipa_table$group <- as.factor(read_ipa_table$group)
ipa_list <- split(read_ipa_table, f = read_ipa_table$group)

comb0<- left_join(read_ipa_table, ipa_list[[4]], by = "pathway")
comb1 <- left_join(comb0, ipa_list[[3]], by = "pathway")
comb2 <- left_join(comb1, ipa_list[[2]], by = "pathway")
comb3 <- left_join(comb2, ipa_list[[1]], by = "pathway")

ipa_table_comb <- unique(comb3[,-c(2,seq(3,11, by=2))])
colnames(ipa_table_comb) <- c("pathway",rev(names(ipa_list)))

ipa_table_comb[is.na(ipa_table_comb)] =0
ord <- hclust( dist(ipa_table_comb, method = "euclidean"), method = "ward.D" )$order
longData <- melt(ipa_table_comb, id = "pathway")
colnames(longData) <- c("pathway", "group", "log_p")
longData$log_p <- as.numeric(longData$log_p)
head(longData, 20)
max(longData$log_p)
min(longData$log_p)

longData$pathway<- factor( longData$pathway, levels = ipa_table_comb$pathway[ord] )
longData$group <- factor(longData$group, levels=c("2vs4","1vs2","3vs4","1vs3"))
myPalette <- colorRampPalette(brewer.pal(9, "YlGnBu"), space="Lab")
require(grid) # for unit()
p=ggplot(longData,aes(x = group, y = pathway, fill = log_p))+
  geom_tile(colour = "white", size=0.5)+
  scale_fill_gradientn(colours = myPalette(100),name=expression("-log"[10]*"(P-value)"),limits = c(min(longData$log_p),max(longData$log_p)),oob=squish, na.value = 'salmon')+
  coord_equal()+
  theme(legend.position="right")+
  theme(text = element_text(size=14), axis.text.y = element_text(size=10),
        axis.text.x = element_text(angle = 60, vjust = 1, size = 10, hjust = 1),
        axis.title.y = element_blank(), axis.title.x = element_blank(), panel.background = element_rect(fill = "white",colour = "white"),  panel.grid.major = element_line(size = 0.5, linetype = 'solid', colour = "black"), panel.grid.minor = element_line(size = 0.25, linetype = 'solid',colour = "black"))

```
